### PR TITLE
24_缪逸伦_第三课作业

### PR DESCRIPTION
**Pull request by Lilmmmm**

### CODE REVIEW

File: 24_缪逸伦_第三课作业 [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+L(K2) = [K2, A, C, O]
+
+L(Z) = [Z] + merge(L(K1), L(K2), [K1, K2])
+	 = [Z] + merge([K1, A, B, O], [K2, A, C, O], [K1, K2])
+	 = [Z, K1] + merge([A, B, O], [K2, A, C, O], [K2])
+	 = [Z, K1, K2] + merge([A, B, O], [A, C, O])
+	 = [Z, K1, K2, A] + merge([B, O], [C, O])
+	 = [Z, K1, K2, A, B] + merge([O], [C, O])
+	 = [Z, K1, K2, A, B, C] + merge([O], [O])
+	 = [Z, K1, K2, A, B, C, O]
```
Comment (jonnyhsy) [inline 24_缪逸伦_第三课作业]: 这个结果不太对，其中注意：contract K1 is A, B {...} 这个意思是：合约K1
继承自合约B，合约A，其中B是最先被继承的合约("most-derived"), 继承链应该是K1 <--- B <--- A 见文档：
 - https://solidity.readthedocs.io/en/develop/contracts.html#inheritance  值得好好读一遍

正确的结果：
L(O) = [O]
L(A) = [A, O]
L(B) = [B, O]
L(C) = [C, O]
L(K2) = [K2] + merge(L(C), L(A), [C, A])
      = [K2] + merge([C, O], [A, O], [C, A])
      = [K2, C] + merge([O], [A, O], [A])
      = [K2, C, A, O]
L(K1) = [K1] + merge(L(B), L(A), [B, A])
      = [K1] + merge([B, O], [A, O], [B, A])
      = [K1, B] + merge([O], [A, O], [A])
      = [K1, B, A, O]
L(Z) = [Z] + merge(L(K2), L(K1), [K2, K1])
     = [Z] + merge([K2, C, A, O], [K1, B, A, O], [K2, K1])
     = [Z, K2] + merge([C, A, O], [K1, B, A, O], [K1])
     = [Z, K2, C] + merge([A, O], [K1, B, A, O], [K1])
     = [Z, K2, C, K1] + merge([A, O], [B, A, O])
     = [Z, K2, C, K1, B] + merge([A, O], [A, O])
     = [Z, K2, C, K1, B, A, O]

===== test.py =====  (Note: python的新式类继承用的就是C3算法，所以可以用来测试下)
```python
class O(object):                                                                   
  def __init__(self):                                                              
    print "O"                                                                      

class A(O):                                                                        
  def __init__(self):                                                              
    print "A"                                                                      

class B(O):                                                                        
  def __init__(self):                                                              
    print "B"                                                                      

class C(O):                                                                        
  def __init__(self):                                                              
    print "C"                                                                      

class K1(B, A):                                                                    
  def __init__(self):                                                              
    print "K1"                                                                     

class K2(C, A):                                                                    
  def __init__(self):                                                              
    print "K2"                                                                     

class Z(K2, K1):                                                                   
  def __init__(self):                                                              
    print "Z"


print Z.__mro__
(<class '__main__.Z'>, <class '__main__.K2'>, <class '__main__.C'>, <class '__main__.K1'>, <class '__main__.B'>, <class '__main__.A'>, <class '__main__.O'>, <type 'object'>)
```

File: 24_缪逸伦_第三课作业 [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    uint constant payDuration = 10 seconds;
+    uint          totalSalary = 0;
+    address       owner;
+    mapping(address => Employee) public employees;
+    
+    function Payroll() {
+        owner = msg.sender;
+    }
+    
+    modifier onlyOwner {
```
Comment (jonnyhsy) [inline 24_缪逸伦_第三课作业]: Note: 如果导入了OpenZepplin库里的Ownable.sol和SafeMath.sol可以重用代码，不需要再定义owner和onlyOwner() modifier.
```
import "github.com/OpenZeppelin/zeppelin-solidity/contracts/math/SafeMath.sol";
import "github.com/OpenZeppelin/zeppelin-solidity/contracts/ownership/Ownable.sol";
```

File: 24_缪逸伦_第三课作业 [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    }
+    
+    modifier employeeExist(address employeeId) {
+        var employee = employees[employeeId];
+        assert(employee.id != 0x0);
+        _;
+    }
+    
+    function _partialPaid(Employee employee) private {
+        uint payment = employee.salary * (now - employee.lastPayday) / payDuration;
```
Comment (jonnyhsy) [inline 24_缪逸伦_第三课作业]: 建议涉及到数学运算的都使用OpenZepplin的SafeMath.sol来防止溢出

Comment (Lilmmmm) [reply 24_缪逸伦_第三课作业]: 已修改

**Pull request merged**